### ISSUE TITLE
rewrite and implement build/test/deploy pipelines for github

### ISSUE THREAD

**Issue by epag** [epag/wresTicketTesting#140]
---

Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Redmine Issue: 126574, https://vlab.noaa.gov/redmine/issues\126574
Original Date: 2024-02-14
Original Assignee: evan.pagryzinski

---

Instead of relying on our Jenkins server, we hope to be able to fully leverage github actions. This ticket is to track the work moving and improve our current infrastructure to github

Useful wikis:
Container registry for docker images: https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry
Github packages for artifacts: https://docs.github.com/en/packages/learn-github-packages/introduction-to-github-packages
Workflows with gradle: https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-gradle-registry
Github secrets: https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions?tool=webui

Action to integrate with Jenkins: https://github.com/marketplace/actions/trigger-jenkins-job

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T14:02:00Z

---

Spending a couple days reading through the github action docs again and our current system to try and come up with a solid plan for migration. Will add an estimate to this ticket after that

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T14:33:36Z

---

Hank or James, out of these projects currently on github, are there any that are similar to architecture like WRES? https://github.com/orgs/NOAA-OWP/repositories

I know we would be able to do the follow:
- when someone submit code we build and run unit tests
- possibly run systests (If we can get that 32g of data accessible in the repo)

What I am trying to find examples of:
- someone interfacing with government systems

Basically I am trying to determine if we will still have to do all of the manual deploy steps and now we are just getting the builds from github or if there is more we can move over

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-02-14T14:37:28Z

---

We can take a look around during tomorrow's call if you want, but I don't think any are quite like the WRES, if only because none of them are for Java-based software. Furthermore, I don't think any have the level of testing that we have or are for services in "production".

If you are looking for examples of Actions, let me know. I know that some are using Actions, I'm just not sure what those Actions are doing.

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T14:46:06Z

---

hank.herr wrote:
> We can take a look around during tomorrow's call if you want, but I don't think any are quite like the WRES, if only because none of them are for Java-based software. Furthermore, I don't think any have the level of testing that we have or are for services in "production".
> 
> If you are looking for examples of Actions, let me know. I know that some are using Actions, I'm just not sure what those Actions are doing.
> 
> Hank

I have found ones that are using actions, just trying to determine the scope of the work we are going to be able to do in github. I think at a baseline we will be able to build and run unit tests, storing the distros somewhere (Need to find a solution/example of this, but see no reason why we wouldn't be able to). We may be able to do more, but trying to figure that out now. 

My worry is something like:
- We aren't able to execute system tests/external system tests on github
- Since Jenkins is on gov hardware we are blocked for kicking off a build on our server (Github workflows should be able to trigger Jenkins stuff, but unsure with security related stuff)

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T15:45:24Z

---

*A couple open questions:*

*- Are we allowed to use open source (And free) github actions?* 
If we are unable to do system and external service tests on github (Likely), do any of you see issue with leveraging an open source github action to trigger a Jenkins build (It could pull down from the github and do its work as normal)? Who would we ask about this?

*- Is our Jenkins server behind the government firewall?*
Could the above action (Assuming we can even use it) be ABLE to access our jenkins server to kick off a build? If not, I need to do some thinking on how to go about doing, at the very least, external service tests

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-02-14T15:48:57Z

---

evan.pagryzinski wrote:
> *A couple open questions:*
> 
> *- Are we allowed to use open source (And free) github actions?* 
> If we are unable to do system and external service tests on github (Likely), do any of you see issue with leveraging an open source github action to trigger a Jenkins build (It could pull down from the github and do its work as normal)? Who would we ask about this?

You would probably ask OWP github admin... i.e., me. And, I would immediately respond, "I have no idea." :)

I'm still on the hook for giving you a legible version of the OWP policy. Maybe that would help.  Let me make a copy and share it.

> *- Is our Jenkins server behind the government firewall?*
> Could the above action (Assuming we can even use it) be ABLE to access our jenkins server to kick off a build? If not, I need to do some thinking on how to go about doing, at the very least, external service tests

No. James can see it, and lives in the England, our nemesis from the Revolutionary War and War of 1812, so it can't be behind a firewall.

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T15:56:02Z

---

hank.herr wrote:
> evan.pagryzinski wrote:
> > *- Is our Jenkins server behind the government firewall?*
> > Could the above action (Assuming we can even use it) be ABLE to access our jenkins server to kick off a build? If not, I need to do some thinking on how to go about doing, at the very least, external service tests
> 
> No. James can see it, and lives in the England, our nemesis from the Revolutionary War and War of 1812, so it can't be behind a firewall.
> 
> Hank

I just tried from my local machine and first it prompted for credentials. I gave it some junk credentials and it then prompted for me to log in with my noaa email/password. There might be some fenagling needed to interface with it. Do you know how we do it present day with vlab? Or is it not an issue because vlab is in the same network? 

We might need some way to login to jenkins with some default user (Personally I don't want to tie it to just my account ya know)

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-02-14T15:57:24Z

---

Evan:

I've attached the NOAA guidelines. The OWP guidelines reuse this text nearly verbatim with the addition of a table spelling out roles and responsibilities. Since I haven't yet seen the final form of that document, I don't know if I should share what I have (a copy from Jeff). However, the NOAA document is final; hence, I'm sharing that. Let me know if this helps. 

As for the use of 3rd party tools, I don't recall it being directly addressed. However, I would see no reason for concern as long as there are no known security issues with that tool.

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **james.d.brown** (james.d.brown)
Original Date: 2024-02-14T15:58:36Z

---

Yeah, but I have special access as your overlord... plus I need to log in.

I think it's clear what we should aim for, namely the full pipeline on github. I agree that there are questions about how much of it is achievable, but we should start by assuming "all" and work out how to do it. Afterall, millions of people worldwide have their entire pipelines on github. We may need to occasionally propagate to jenkins to satisfy some internal requirements, but we should do it only insofar as that is true.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **james.d.brown** (james.d.brown)
Original Date: 2024-02-14T16:02:05Z

---

evan.pagryzinski wrote:
> we will be storing artifacts (Should be under the size restraint of 500MB)

I think that size limit is for private repos, no? Or is this a separate limit for artifacts attached to public projects? Either way, we probably need to add an artifact repo into our pipeline. Whether there are free ones, idk.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **james.d.brown** (james.d.brown)
Original Date: 2024-02-14T16:03:37Z

---

By artifact repo, I mean something like artifactory or nexus, but these are commercial services.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-02-14T16:04:11Z

---

> I just tried from my local machine and first it prompted for credentials. I gave it some junk credentials and it then prompted for me to log in with my noaa email/password. There might be some fenagling needed to interface with it. Do you know how we do it present day with vlab? Or is it not an issue because vlab is in the same network? 
> 
> We might need some way to login to jenkins with some default user (Personally I don't want to tie it to just my account ya know)

While our overlord has been granted a username and password, you and I, Evan, as pleebs, need to use your CAC cards. :) 

The @wres-cron@ account logs in to obtain artifacts using a special @wres-admin@ account, iirc. You can see how it works by logging into @nwcal-wres-ti01@, sudo-ing to @wres-cron@ and checking out the install scripts in /wres_share/releases/install_scripts/. Exact directory name may be off. 

But I agree with James that we should try to make GitHub do as much as possible. Hopefully, that means at least building the artifacts. If push comes to shove, we can setup system testing as we do it now, though that would be suboptimal (but probably unavoidable given the size of our data sets).  

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-02-14T16:04:54Z

---

james.d.brown wrote:
> evan.pagryzinski wrote:
> > we will be storing artifacts (Should be under the size restraint of 500MB)
> 
> I think that size limit is for private repos, no? Or is this a separate limit for artifacts attached to public projects? Either way, we probably need to add an artifact repo into our pipeline. Whether there are free ones, idk.

Yes, I believe that's correct. I can double check, but I think public has unlimited size. 

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T16:08:53Z

---

hank.herr wrote:
> james.d.brown wrote:
> > evan.pagryzinski wrote:
> > > we will be storing artifacts (Should be under the size restraint of 500MB)
> > 
> > I think that size limit is for private repos, no? Or is this a separate limit for artifacts attached to public projects? Either way, we probably need to add an artifact repo into our pipeline. Whether there are free ones, idk.
> 
> Yes, I believe that's correct. I can double check, but I think public has unlimited size. 
> 
> Hank

The repo itself has unlimited size, but there are limitation on the size of data produced by github actions/workflows even for public repos https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#included-storage-and-minutes

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T16:10:33Z

---

Oh wait, I had misinterpreted the above, that is the amount of storage on github free for private repos, my mistake, we should be good on size since it is a public repo

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **james.d.brown** (james.d.brown)
Original Date: 2024-02-14T16:17:04Z

---

> You can login without a cac too, just use your email credentials :) It's just easier with cac.

Oh yeah. I've only ever used CAC, so I forgot there was an alternative.

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **hank.herr** (hank.herr)
Original Date: 2024-02-14T16:23:16Z

---

Deleted the accidental relation. My bad,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T19:23:49Z

---

In regards to system tests: 

Still reading, but it seems like there is no official limitation (At least one written) on the size of a repository. *However*, @We recommend repositories remain small, ideally less than 1 GB, and less than 5 GB is strongly recommended@. It seems like we could POTENTIALLY make a sub repo of our WRES repo that contains system testing data and check that out to do system tests in the workflow, but github makes a hand-wavy statement saying @If your repository excessively impacts our infrastructure, you might receive an email from GitHub Support asking you to take corrective action@

So seems like we could try to do this, but also we should still keep the jenkins pipline intact incase github complains

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **james.d.brown** (james.d.brown)
Original Date: 2024-02-14T19:47:39Z

---

I don't think we'd want to host these datasets in a code repo. We don't even do that now, except for the "smalldata" (edit: which we also shouldn't really do). As with the artifacts we produce on the backend, we'd want to store these in an artifact/binary repo. As to what repo and what the pipeline looks like (and how performant it is), idk. But at the point it is not performant, we may want to revisit some of these datasets for system testing, notably the nwm datasets. It was always a compromise and we favored retaining the "production-like" nwm data, rather than slicing it, but we may need to rethink that.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **evan.pagryzinski** (evan.pagryzinski)
Original Date: 2024-02-14T19:53:45Z

---

james.d.brown wrote:
> I don't think we'd want to host these datasets in a code repo. We don't even do that now, except for the "smalldata" (edit: which we also shouldn't really do). As with the artifacts we produce on the backend, we'd want to store these in an artifact/binary repo. As to what repo and what the pipeline looks like (and how performant it is), idk. But at the point it is not performant, we may want to revisit some of these datasets for system testing, notably the nwm datasets. It was always a compromise and we favored retaining the "production-like" nwm data, rather than slicing it, but we may need to rethink that.

Yeah, I thought it was bad smell. I am currently just trying to go under the assumption of "Move everything to github and find potential problems"
